src/state.py

```diff
@@ -215,25 +215,74 @@ class _CloudsAuthConfig(typing.TypedDict):
         username: The OpenStack user name for given project.
     """
 
+    # This is a configuration class and does not have methods.
+    class Config:  # pylint: disable=too-few-public-methods
+        """Pydantic configuration options.
+
+        Attributes:
+            frozen: Define whether the object should be mutable.
+        """
+
+        frozen = True
+
     auth_url: str
     password: str
     project_domain_name: str
     project_name: str
     user_domain_name: str
     username: str
 
+    def get_id(self) -> str:
+        """Get unique cloud configuration ID.
+
+        Returns:
+            The unique cloud configuration ID.
+        """
+        return f"{self.project_name}_{self.username}"
+
+    @classmethod
+    def from_unit_relation_data(cls, data: ops.RelationDataContent) -> "CloudsAuthConfig | None":
+        """Get auth data from unit relation data.
+
+        Args:
+            data: The unit relation data.
+
+        Returns:
+            CloudsAuthConfig if all required relation data are available, None otherwise.
+        """
+        if any(
```

Comment:
Perhaps log this?

Reply:
Added info log, thank you for the review!